### DISCUSSION TITLE
2023-04-20 monthly meeting

### DISCUSSION THREAD

**Discussion by explodecomputer** [MRCIEU/opengwas#30] [Meeting notes]
Agenda

- [ ] Yi trying GPT-4 for trait mapping - **aim for next month**
- [ ] Data management interface **discussed implementing QC process see below**
- [ ] Discuss new time so that Andrew can join **I'll change to 3pm start going forwards**
- [ ] CI/CD **Andrew to look at this and recommend how to proceed**
- [ ] New data uploads **Gib to deal with**

## Data QC check during upload

1. Finalise integration and merge into main branch @mightyphil2000; include tests on dummy data to make sure errors are matched @explodecomputer 
2. Update API to accept new QC info @JasonQiu @explodecomputer 
3. Update GwasDataImport to also upload QC info @JasonQiu @explodecomputer 
4. Update HPC method to include QC step @explodecomputer 
5. Update opengwas-reports to include new QC info @JasonQiu @YiLiu6240 @explodecomputer 

## CI/CD

- Some progress made by Tessa on CI but CD not complete
- Good to be able to have easy deployment that isn't dependent on a single person
- But automated deployment might be risky
- Andrew to look at current CI/CD state and suggest future directions

## Load issues

- Rita raised that often usage is too high and API times out
- restricting by IP to some number of queries e.g. 500
- avoid blocking UoB server access
- consider no restriction for authenticated users
- consider adding a different oauth2 provider to avoid excluding some international users

**Comment by remlapmot** [comment]
- [ ] I can show that I made some updates to the **TwoSampleMR** site <https://mrcieu.github.io/TwoSampleMR/>, e.g. added Search facility
- [ ] There were a few dead URLs, e.g., where did <https://github.com/MRCIEU/ieugwasr/blob/master/README.md#authentication> move to?
- [ ] Also the **gwasglue** site could do with a refresh <https://mrcieu.github.io/gwasglue/>
- [ ] **ieugwasr** has a site at <https://mrcieu.github.io/ieugwasr/>, the site needs refreshing and the URL can be added its repo **About** section

**Reply by explodecomputer** [reply]
Many thanks @remlapmot! 
This is the new location for the authentication info https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication-1

**Reply by explodecomputer** [reply]
added the ieugwast site to About section but will make an issue for the site refresh

**Reply by remlapmot** [reply]
Thanks - I have added that back into the 3 vignettes (introduction.Rmd, exposure.Rmd, and outcome.Rmd) - sorry I missed it was actually in gwas2020.Rmd.

**Comment by mightyphil2000** [comment]
- [ ] I've identified 18 datasets in Open GWAS with probable effect allele metadata errors. 9 of the 18 are from the EBI batch. This is an error rate of 1.15% (the analysis excluded the prot, eqtl and ukb batches). Metadata errors were flagged by comparing known GWAS hits for each trait in Open GWAS with the GWAS catalog and that had >10 SNPs available for the comparison. However, no filter on sample size was used. Using a more relaxed threshold, I identified 213 datasets with a possible effect allele metadata error.  

 [1] "ebi-a-GCST000755" "ebi-a-GCST000759" "ebi-a-GCST000760" "ebi-a-GCST001592"
 [5] "ebi-a-GCST004772" "ebi-a-GCST006305" "ebi-a-GCST008592" "ebi-a-GCST008593"
 [9] "ebi-a-GCST008594" "finn-b-J10_COPD"  "ieu-a-1013"       "ieu-a-1050"      
[13] "ieu-a-818"        "ieu-b-106"        "met-a-428"        "met-a-532"       
[17] "met-c-850"        "met-d-Creatinine"

**Reply by explodecomputer** [reply]
Really helpful thanks @mightyphil2000 . It's strange that met-a, met-d and met-c are in there - as these are batch uploads which should have consistent effect alleles. Any thoughts on what could be going on there?

**Reply by mightyphil2000** [reply]
not sure. I will check further.

**Reply by mightyphil2000** [reply]
> Really helpful thanks @mightyphil2000 . It's strange that met-a, met-d and met-c are in there - as these are batch uploads which should have consistent effect alleles. Any thoughts on what could be going on there?

Hey Gib so I've dug a bit further into this. so I think the flagged met-a, met-d and met-c studies are actually ok. Two of the flagged studies have weird trait names (of the sort x-1214), which weren't meaningful. met-c-850 and met-d-Creatinine are both studies of creatinine and I got the same pattern of conflicts with the GWAS catalog. It turns out that a subset of the results in the GWAS catalog correspond to Glomerular filtration rate (creatinine). However, glomerular filtration rate and creatinine are inversely correlated. This created the conflicts with the creatine studies in Open GWAS. 

However, I'm still quite worried about the ebi batch. Most of the conflicts seem to stem from EBI. A couple of months ago someone did contact me about a problem with one ebi study that turned out to be an effect allele coding error, which I asked Jason to remove. This is what prompted me to do another scan of OpenGWAS. 

Below is an example of another clear effect allele meta data error, corresponding to ebi-a-GCST000755 (HDL cholesterol) identified from my recent search. This plot shows GWAS hits for HDL cholesterol in all GWAS catalog European studies being compared to ebi-a-GCST000755 (HDL cholesterol). 

![image](https://github.com/MRCIEU/opengwas/assets/10760254/1f921c7a-6601-45ac-a4e9-a3871ce9d27d)

This next plot compares GWAS hits for HDL cholesterol between the GWAS catalog and Open GWAS but restricted to the same study (ie ebi-a-GCST000755). In other words, the same two studies are being compared and so there can be no doubt that there is an effect allele metadata error. 
![image](https://github.com/MRCIEU/opengwas/assets/10760254/4bb33b54-db53-46f0-aa84-c0e0985d49fd)

It seems to me that the effect allele column is not being consistently identified correctly in the datasets being imported from the GWAS catalog.

**Comment by mightyphil2000** [comment]
- [ ] My UK Biobank application is going to be retired by February 2025. I think the OpenGWAS datasets generated in UK Biobank are currently done under my application? This process will eventually have to transfer to Lavinia's UKB application.

**Reply by explodecomputer** [reply]
What are the implications for this? If they aren't transferred do they need to be taken down or something?

**Reply by mightyphil2000** [reply]
I don't think they will need to be taken down. I will followup with UK Biobank to clarify that.

**Comment by remlapmot** [comment]
I have made some fixes to GwasDataImport after the integration with CheckSumStats and fixed its pkgdown site <https://mrcieu.github.io/GwasDataImport/>.

Some points for chatting about at the next meeting.

* Could we stop pinning the version of Bioconductor in our R package DESCRIPTION files - I removed the 3.11 in the DESCRIPTION file of GwasDataImport. Otherwise when installing MRCIEU R packages we and users can get caught in frustrating dependency cycles when installing packages, e.g., we specified Bioc 3.11 versions of Bioc packages in one R package and then the current version of Bioc packages in another of our R packages - leading to re-installations of different versions of the same Bioc packages ...
* We need to start doing versioning more correctly - as we only distribute packages through GitHub we really should bump minor/patch version numbers every time that new code is pushed/merged to the master/main branch. Hence I moved GwasDataImport to 0.1.2 and git tagged it v0.1.2. I feel bad that there is currently so much additional code on TwoSampleMR master after the v0.5.6 tag.

**Comment by remlapmot** [comment]
Just noting that I have had to make my commits into TwoSampleMR very late at night just so the OpenGWAS server isn't maxed out so the tests will run. And I think that @ritarasteiro was also having problems getting tests to run during the day. And here is another user having problems <https://github.com/jean997/cause/issues/36>.

**Comment by JasonQiu** [comment]
Just wanted to clarify if below is the (entire and only) QC info that should be uploaded to API?

e.g.
```shell
zv21942@ieu-db-interface:~$ cat /data/igd/ieu-b-5089/qc_metrics.json
{"af_correlation":0.9503,"inflation_factor":1.1474,"mean_EFFECT":-0,"n":"-Inf","n_snps":12321875,"n_clumped_hits":82,"n_p_sig":14022,"n_mono":0,"n_ns":1293897,"n_mac":0,"is_snpid_unique":false,"n_miss_EFFECT":0,"n_miss_SE":0,"n_miss_PVAL":0,"n_miss_AF":0,"n_miss_AF_reference":1423040,"n_est":"NA","ratio_se_n":"NA","mean_diff":"NaN","ratio_diff":"NaN","sd_y_est1":"NaN","sd_y_est2":"NA","r2_sum1":0,"r2_sum2":0,"r2_sum3":0,"r2_sum4":0,"ldsc_nsnp_merge_refpanel_ld":0,"ldsc_nsnp_merge_regression_ld":"NA","ldsc_observed_scale_h2_beta":"NA","ldsc_observed_scale_h2_se":"NA","ldsc_intercept_beta":"NA","ldsc_intercept_se":"NA","ldsc_lambda_gc":"NA","ldsc_mean_chisq":"NA","ldsc_ratio":"NA"}
```

If yes then it looks like we need to set up a new type of node `QCResult` and a new relationship `HAS_QC` such that `(:GwasInfo)-[:HAS_QC]->(:QCResult)`.

Can I assume that if QC has been done multiple times on a same dataset then the latest result will always overwrite?

When and how shall the QC info be uploaded to Neo4j? To avoid modifying Cromwell and Docker configurations, this can be done within the `/quality_control/release` endpoint, when [DID_QC relationship is added](https://github.com/MRCIEU/opengwas-api/blob/53b3a7ee0ac987a5df6c75c387562c77d51b628e/app/apis/quality_control.py#L90). The content of the JSON file will be  used to create the `QCResult` node, before the relationship `HAS_QC` is set up. This will not happen automatically when the QC process is completed - it will be triggered manually, usually by calling `ieugwasr::api_gwas_release()`.

Alternatively, I think there should be a kind of 'hook' mechanism in Cromwell so that an additional API endpoint (solely built for uploading qc_metrics.json) can be called automatically when the QC job finishes. This requires more work and may bring uncertainty.

**Reply by explodecomputer** [reply]
There are two endpoints for uploading -
- metadata: https://github.com/MRCIEU/opengwas-api/blob/master/app/apis/edit.py#L40
- gwas summary stats file: https://github.com/MRCIEU/opengwas-api/blob/master/app/apis/edit.py#L212

Perhaps easiest to add a third endpoint for uploading the qc statistics object. e.g. `upload_qc`. Within that endpoint it would update neo4j with the relevant information.

Adding the `QCResult` node and `HAS_QC` relationship makes sense to me. We'd have to think a bit about what is stored in that node.

Agree that we only keep the latest.
